File: HNQQPopMenu-OC/HNQQPopMenu/HNQQPopmenu/HNQQPopMenuManager.m
//
//  HNQQPopMenuManager.m
//  HNQQPopMenu
//
//  Created by 许浩男 on 16/4/2.
//  Copyright © 2016年 Zakariyya. All rights reserved.
//

#import "HNQQPopMenuManager.h"

@interface HNQQPopMenuManager()

@property (nonatomic,strong) HNQQPopMenuView *popmenuView;

@end

@implementation HNQQPopMenuManager

+ (instancetype)sharedManager{
    static HNQQPopMenuManager *manager;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        manager = [[HNQQPopMenuManager alloc] init];
    });
    return manager;
}

+ (void)showPopMenuWithView:(UIView *)view items:(NSArray *)itemArr delegate:(id<HNQQPopMenuViewDelegate>)delegate dismissAutomatically:(BOOL)dismissed{
    [[self sharedManager] showPopMenuWithView:view items:itemArr delegate:delegate dismissedAutomatically:dismissed];
}

+ (void)showPopMenuWithView:(UIView *)view items:(NSArray *)itemArr action:(action)action dismissAutomatically:(BOOL)dismissed{
    [[self sharedManager] showPopMenuWithView:view items:itemArr action:action dismissedAutomatically:dismissed];
}

+ (void)setBackgroundColor:(UIColor *)color{
    [[self sharedManager] setBackgroundColour:color];
}

+ (void)setBackgroundView:(UIView *)view{
    [[self sharedManager] setBackView:view];
}

+ (void)dismiss{
    [[self sharedManager] dismiss];
}


- (instancetype)init{
    [NSException raise:NSInvalidArgumentException format:@"%s Using the %@ initializer directly is not supported. Use %@ or %@ instead.", __PRETTY_FUNCTION__,NSStringFromSelector(@selector(init)),NSStringFromSelector(@selector(showPopMenuWithView:items:action:dismissedAutomatically:)),NSStringFromSelector(@selector(showPopMenuWithView:items:delegate:dismissedAutomatically:))];
    return nil;
}

#pragma mark - implementation
- (void)showPopMenuWithView:(UIView *)view items:(NSArray *)itemArr{
    if (self.popmenuView) {
        [self dismiss];
    }
    self.popmenuView = [[HNQQPopMenuView alloc] initWithView:view items:itemArr];
    self.popmenuView.frame = [UIScreen mainScreen].bounds;
    [[UIApplication sharedApplication].keyWindow addSubview:self.popmenuView];
    [UIView animateWithDuration:0.25f animations:^{
            self.popmenuView.tableView.transform = CGAffineTransformMakeScale(1.0, 1.0);
    }];
    
}

- (void)showPopMenuWithView:(UIView *)view items:(NSArray *)itemArr delegate:(id<HNQQPopMenuViewDelegate>)delegate dismissedAutomatically:(BOOL)dismissed{
    [self showPopMenuWithView:view items:itemArr];
    self.popmenuView.delegate = delegate;
    self.popmenuView.dismissed = dismissed;
}

- (void)showPopMenuWithView:(UIView *)view items:(NSArray *)itemArr action:(action)action dismissedAutomatically:(BOOL)dismissed{
    [self showPopMenuWithView:view items:itemArr];
    self.popmenuView.dismissed = dismissed;
    if (action) {
        self.popmenuView.clickAction = action;
    }
}

- (void)setBackgroundColour:(UIColor *)color{
    if (self.popmenuView) {
        self.popmenuView.tableView.backgroundColor = color;
    }
}

- (void)setBackView:(UIView *)view{
    if (self.popmenuView) {
        self.popmenuView.tableView.backgroundView = view;
    }
}

- (void)dismiss{
    [UIView animateWithDuration:0.2f animations:^{
        self.popmenuView.tableView.transform = CGAffineTransformMakeScale(0.001, 0.001);
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.1f * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
            [self.popmenuView.triangleLayer removeFromSuperlayer];
        });
    } completion:^(BOOL finished) {
        [self.popmenuView.tableView removeFromSuperview];
        [self.popmenuView removeFromSuperview];
        self.popmenuView = nil;
    }];
}



@end

File: HNQQPopMenu-swift/HNQQPopMenu-swift/HNQQPopMenu/HNQQPopMenuManager.swift
//
//  HNQQPopMenuManager.swift
//  HNQQPopMenu-swift
//
//  Created by zakariyyaSv on 16/4/22.
//  Copyright © 2016年 Zakariyya. All rights reserved.
//

import UIKit

typealias actionClosure = (NSInteger) -> Void


class HNQQPopMenuManager: NSObject {
    
    private var popMenuView : HNQQPopMenuView!
    
    private class var sharedInstance: HNQQPopMenuManager {
        struct Static {
            static var once :dispatch_once_t = 0
            static var instance : HNQQPopMenuManager? = nil
        }
        dispatch_once(&Static.once){
            Static.instance = HNQQPopMenuManager()
        }
        
        return Static.instance!
    }
    
    class func showPopMenuView(view:UIView,itemArr:Array<HNQQPopMenuModel>,delegate:HNQQPopMenuViewDelegate,dismissAutomatically:Bool){
        self.sharedInstance.showPopMenuView(view, itemArr: itemArr, delegate: delegate, dismissedAutomatically: dismissAutomatically)
    }
    
    class func showPopMenuView(view:UIView,itemArr:Array<HNQQPopMenuModel>,action:actionClosure,dismissAutomatically:Bool){
        self.sharedInstance.showPopMenuView(view, itemArr: itemArr, action: action, dismissedAutomatically: dismissAutomatically)
    }
    
    class func setBackgroundColor(color:UIColor){
        self.sharedInstance.setBackgroundColor(color)
    }
    
    class func setBackgroundView(view:UIView){
        self.sharedInstance.setBackView(view)
    }
    
    class func dismiss(){
       self.sharedInstance.dismiss()
    }
    
    
    private func showPopMenuView(view:UIView,itemArr:Array<HNQQPopMenuModel>,delegate:HNQQPopMenuViewDelegate,dismissedAutomatically:Bool){
        self.showPopMenuView(view, itemArr: itemArr)
        popMenuView.delegate = delegate
        popMenuView.dismissed = dismissedAutomatically
    }
    
    private func showPopMenuView(view:UIView,itemArr:Array<HNQQPopMenuModel>,action:actionClosure?,dismissedAutomatically:Bool){
        self.showPopMenuView(view, itemArr: itemArr)
        if (action != nil) {
            popMenuView.action = action
        }
        popMenuView.dismissed = dismissedAutomatically
    }
    
    private func showPopMenuView(view:UIView,itemArr:Array<HNQQPopMenuModel>){
        if (popMenuView != nil) {
            self.dismiss()
        }
        popMenuView = HNQQPopMenuView.init(view: view, itemArr: itemArr)
        popMenuView.frame = UIScreen.mainScreen().bounds
        UIApplication.sharedApplication().keyWindow?.addSubview(popMenuView)
        UIView.animateWithDuration(0.25) { 
            () -> Void in
            self.popMenuView.tableView.transform = CGAffineTransformMakeScale(1.0, 1.0)
        }
    }
    
    private func setBackgroundColor(color:UIColor){
        if (popMenuView != nil) {
            popMenuView.tableViewBackgroundColor(color)
        }
    }
    
    private func setBackView(view:UIView){
        if (popMenuView != nil) {
            popMenuView.tableView.backgroundView = view
        }
    }
    
    private func dismiss(){
        
        UIView.animateWithDuration(0.2, animations: { 
            () -> Void in
            self.popMenuView.tableView.transform = CGAffineTransformMakeScale(0.001, 0.001)
            let time = dispatch_time(DISPATCH_TIME_NOW,(Int64)(100 * NSEC_PER_MSEC))
            dispatch_after(time, dispatch_get_main_queue(), {
                () -> Void in
                self.popMenuView?.triangleLayer?.removeFromSuperlayer()
            })

            }) { (completed) in
                self.popMenuView?.tableView.removeFromSuperview()
                self.popMenuView.removeFromSuperview()
                self.popMenuView = nil
        }
    }
    
}
